File: 백준/Gold/2470. 두 용액/두 용액.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;


    public static void main(String[] args) throws Exception {

        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }


        Arrays.sort(arr);


        int left = 0;
        int right = n - 1;

        int result = Integer.MAX_VALUE;
        int[] ans = new int[2];


        while (left < right) {
            int s_left = arr[left];
            int s_right = arr[right];

            int mix = s_left + s_right;

            if (Math.abs(mix) < result) {
                result = Math.abs(mix);
                ans[0] = s_left;
                ans[1] = s_right;
            }


            if (mix < 0) left++;
            else right--;

        }

        bw.write(ans[0] + " " + ans[1]);
        bw.close();
    }
}

File: 백준/Gold/2470. 두 용액/두 용액.py
import sys

input = lambda: sys.stdin.readline().rstrip()


N = int(input())

arr = list(map(int, input().split()))

arr.sort()


left = 0
right = len(arr) - 1

result = []
min_val = float("inf")

while left < right:
    minus = arr[right] + arr[left]

    if minus == 0:
        result = [arr[left], arr[right]]
        min_val = 0
        break
    if min_val > abs(minus):
        min_val = abs(minus)
        result = [arr[left], arr[right]]
        if result == 0:
            break

    if minus < 0:
        left += 1
    else:
        right -= 1


print(*result)
